index.js

```diff
@@ -320,6 +326,93 @@ module.exports = {
     return options;
   },
 
+  _shouldHandleTypeScript() {
+    if (this._cachedShouldHandleTypescript === undefined) {
+      // TODO: this doesn't account for things like in-repo addons and other
+      // weird inclusion structures
+      let checker = new VersionChecker(this.parent).for('ember-cli-typescript', 'npm');
+
+      this._cachedShouldHandleTypescript = checker.gte('4.0.0-alpha.0');
+    }
+
+    return this._cachedShouldHandleTypescript;
+  },
+
+	_buildClassFeaturePluginConstraints(constraints) {
+    // With older versions of ember-cli-typescript, class feature plugins like
+    // @babel/plugin-proposal-class-properties were run before the TS transform.
+    // With more recent language features like `declare` field modifiers, Babel
+    // now has assertions requiring that the TS transform runs first, so if we
+    // know we're responsible for setting that transform up, we follow those rules.
+    if (this._shouldHandleTypeScript()) {
+      constraints.after = constraints.after || [];
```

Comment:
I don't think we need before/after at all since we now control the full list of plugins (we would still need to deal with that in the `else` below though for ember-cli-typescript < 4 support).

Reply:
🤔 yeah, I suppose we only have to force a different order for the exception case (older e-c-ts) since we're just adding them in the correct order to begin with.

Reply:
But what about when something *else* has added `@babel/plugin-transform-typescript`? This PR handles this case (but warns) similar to what `ember-cli-babel` does with `@babel/plugin-proposal-decorators` and `@babel/plugin-proposal-class-properties`. I mean, theoretically, if something else has already added `@babel/plugin-transform-typescript`, we'd be adding the class feature plugins after anyways, but, IMHO, it doesn't hurt to be sure.

Reply:
> But what about when something else has added @babel/plugin-transform-typescript?

1) AFAICT we don't actually have that problem (a search in emberobserver suggests that only the e-c-ts addon does this)
2) I'd rather the code for the _go forward_ world be simpler and clearer. That way when we do a 8.x (or whatever) dropping support for checking for e-c-ts the cleanup is easy.

Reply:
1. I guess I was thinking either something non-Ember or user config.
2. This makes sense. Default to the happy path and just handle the one known issue (pre 4.0 e-c-ts). Anything else is not a supported config (if you're doing something weird, expect breakage).

Reply:
I've changed this to only handle the else case (e-c-ts < 4.0). We still need to enforce the decorators before class properties constraint for now because otherwise they get out of order when the class properties plugin is forced to be *before* the ts transform. Once e-c-babel drops support for e-c-ts < 4.0, the calls to `_buildClassFeaturePluginConstraints()` (including the decorators/class properties before/after) can just be removed altogether.